// @strapi/admin/dist/admin/CreateActionEE-uqyKLxwI.js
'use strict';

Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });

const jsxRuntime = require('react/jsx-runtime');
const React = require('react');
const designSystem = require('@strapi/design-system');
const icons = require('@strapi/icons');
const isNil = require('lodash/isNil');
const reactIntl = require('react-intl');
const useLicenseLimits = require('./useLicenseLimits-FtmLXkwK.js');

const _interopDefault = e => e && e.__esModule ? e : { default: e };

function _interopNamespace(e) {
  if (e && e.__esModule) return e;
  const n = Object.create(null, { [Symbol.toStringTag]: { value: 'Module' } });
  if (e) {
    for (const k in e) {
      if (k !== 'default') {
        const d = Object.getOwnPropertyDescriptor(e, k);
        Object.defineProperty(n, k, d.get ? d : {
          enumerable: true,
          get: () => e[k]
        });
      }
    }
  }
  n.default = e;
  return Object.freeze(n);
}

const React__namespace = /*#__PURE__*/_interopNamespace(React);
const isNil__default = /*#__PURE__*/_interopDefault(isNil);

const CreateActionEE = React__namespace.forwardRef(
  (props, ref) => {
    const { formatMessage } = reactIntl.useIntl();
    const { license, isError, isLoading } = useLicenseLimits.useLicenseLimits();
    const { permittedSeats, shouldStopCreate } = license ?? {};
    if (isError || isLoading) {
      return null;
    }
    return /* @__PURE__ */ jsxRuntime.jsxs(designSystem.Flex, { gap: 2, children: [
      !isNil__default.default(permittedSeats) && shouldStopCreate && /* @__PURE__ */ jsxRuntime.jsx(
        designSystem.Tooltip,
        {
          label: formatMessage({
            id: "Settings.application.admin-seats.at-limit-tooltip",
            defaultMessage: "At limit: add seats to invite more users"
          }),
          side: "left",
          children: /* @__PURE__ */ jsxRuntime.jsx(icons.WarningCircle, { width: "1.4rem", height: "1.4rem", fill: "danger500" })
        }
      ),
      /* @__PURE__ */ jsxRuntime.jsx(
        designSystem.Button,
        {
          ref,
          "data-testid": "create-user-button",
          startIcon: /* @__PURE__ */ jsxRuntime.jsx(icons.Mail, {}),
          size: "S",
          disabled: shouldStopCreate,
          ...props,
          children: formatMessage({
            id: "Settings.permissions.users.create",
            defaultMessage: "Invite new user"
          })
        }
      )
    ] });
  }
);

exports.CreateActionEE = CreateActionEE;
