// @strapi/admin/dist/admin/index.js
'use strict';

Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });

const index = require('./index-YSRONKO4.js');
const Theme = require('./Theme-7W0we6BI.js');
const admin = require('./admin-p0djr5G6.js');



exports.BackButton = index.BackButton;
exports.Blocker = index.Blocker;
exports.ConfirmDialog = index.ConfirmDialog;
exports.DescriptionComponentRenderer = index.DescriptionComponentRenderer;
exports.Filters = index.Filters;
exports.Form = index.Form;
exports.InputRenderer = index.MemoizedInputRenderer;
exports.Layouts = index.Layouts;
exports.Pagination = index.Pagination;
exports.SearchInput = index.SearchInput;
exports.Table = index.Table;
exports.getYupValidationErrors = index.getYupValidationErrors;
exports.renderAdmin = index.renderAdmin;
exports.translatedErrors = index.errorsTrads;
exports.useAdminUsers = index.useAdminUsers;
exports.useFetchClient = index.useFetchClient;
exports.useField = index.useField;
exports.useFocusInputField = index.useFocusInputField;
exports.useForm = index.useForm;
exports.useGuidedTour = index.useGuidedTour;
exports.useInjectReducer = index.useInjectReducer;
exports.useQueryParams = index.useQueryParams;
exports.useTable = index.useTable;
exports.NotificationsProvider = Theme.NotificationsProvider;
exports.Page = Theme.Page;
exports.createContext = Theme.createContext;
exports.useAPIErrorHandler = Theme.useAPIErrorHandler;
exports.useAppInfo = Theme.useAppInfo;
exports.useAuth = Theme.useAuth;
exports.useNotification = Theme.useNotification;
exports.useRBAC = Theme.useRBAC;
exports.useStrapiApp = Theme.useStrapiApp;
exports.useTracking = Theme.useTracking;
exports.FetchError = admin.FetchError;
exports.adminApi = admin.adminApi;
exports.fetchBaseQuery = admin.fetchBaseQuery;
exports.getFetchClient = admin.getFetchClient;
exports.isBaseQueryError = admin.isBaseQueryError;
exports.isFetchError = admin.isFetchError;
